benchmarks/time/src/main/scala/strawman/collection/immutable/HashSetBenchmark.scala

```diff
@@ -15,55 +15,67 @@ import scala.Predef.intWrapper
 @Measurement(iterations = 12)
 @State(Scope.Benchmark)
 class HashSetBenchmark {
-
-  @Param(scala.Array("0", "1", "2", "3", "4", "7", "8", "15", "16", "17", "39", "282", "73121", "7312102"))
+  @Param(scala.Array("0", "1", "2", "3", "4", "7", "8", "15", "16", "17", "39", "282", "4096", "131070", "7312102"))
   var size: Int = _
 
   var xs: HashSet[Long] = _
-  var xss: scala.Array[HashSet[Long]] = _
+  var zs: HashSet[Long] = _
   var zipped: HashSet[(Long, Long)] = _
   var randomIndices: scala.Array[Int] = _
-  var randomIndices2: scala.Array[Int] = _
-  var randomXss: scala.Array[HashSet[Long]] = _
+  def fresh(n: Int) = HashSet((1 to n).map(_.toLong): _*)
 
   @Setup(Level.Trial)
-  def initData(): Unit = {
-    def freshCollection() = HashSet((1 to size).map(_.toLong): _*)
-    xs = freshCollection()
-    xss = scala.Array.fill(1000)(freshCollection())
-    zipped = xs.map(x => (x, x))
+  def initTrial(): Unit = {
     if (size > 0) {
       randomIndices = scala.Array.fill(1000)(scala.util.Random.nextInt(size))
-      randomIndices2 = scala.Array.fill(1000)(scala.util.Random.nextInt(size))
-      randomXss = scala.Array.fill(1000)(freshCollection().take(scala.util.Random.nextInt(size)))
     }
   }
 
+  @Setup(Level.Invocation)
+  def initInvocation(): Unit = {
+    xs = fresh(size)
+    zs = fresh((size / 1000) max 2).map(-_)
+    zipped = xs.map(x => (x, x))
+  }
+
   @Benchmark
-  def incl(bh: Blackhole): Unit = {
-    var ys = HashSet.empty[Long]
+  @OperationsPerInvocation(100)
+  def create(bh: Blackhole): Unit = {
     var i = 0L
-    while (i < size) {
-      ys += i
+    while (i < 100) {
+      bh.consume(fresh(size))
       i += 1
     }
-    bh.consume(ys)
   }
 
   @Benchmark
-  def concat(bh: Blackhole): Unit = bh.consume(xs ++ xs)
-
-  @Benchmark
-  def tail(bh: Blackhole): Unit = bh.consume(xs.tail)
+  @OperationsPerInvocation(1000)
+  def expand_incl(bh: Blackhole): Unit = {
+    var ys = xs
+    var i = 0L
+    while (i < 1000) {
+      ys += -i
+      i += 1
+    }
+    bh.consume(ys)
+  }
```

Comment:
Since we are redesigning the benchmarks I was wondering: de we really need a loop here? Why not the following?

~~~ scala
@Benchmark
def expand_incl(bh: Blackhole): Unit = bh.consume(xs += 42)
~~~

Reply:
The reason for using a loop is to make the benchmark method invocation take more time relative to the time taken by the `initInvocation`-method (which avoids the latter overshadowing the former).

Reply:
Are you sure the `@Setup` time is part of the `@Benchmark` time?

Reply:
No, generally it is not, but as stated in the JavaDoc for `Level.Invocation` (complete text below) benchmark methods taking too little time will skew the result:
>This level is only usable for benchmarks taking more than a millisecond per single benchmark method invocation. It is a good idea to validate the impact for your case on ad-hoc basis as well.
>
>WARNING 1: Since we have to subtract the setup/teardown costs from the benchmark time, on this level, we have to timestamp *each* benchmark invocation. If the benchmarked method is small, then we saturate the system with timestamp requests, which introduce artificial latency, throughput, and scalability bottlenecks.


Complete JavaDoc for `Level.Invocation`:
> Invocation level: to be executed for each benchmark method execution.
>
> WARNING: HERE BE DRAGONS! THIS IS A SHARP TOOL.
> MAKE SURE YOU UNDERSTAND THE REASONING AND THE IMPLICATIONS OF THE WARNINGS BELOW BEFORE EVEN CONSIDERING USING THIS LEVEL.
>
>This level is only usable for benchmarks taking more than a millisecond per single benchmark method invocation. It is a good idea to validate the impact for your case on ad-hoc basis as well.
>
>WARNING 1: Since we have to subtract the setup/teardown costs from the benchmark time, on this level, we have to timestamp *each* benchmark invocation. If the benchmarked method is small, then we saturate the system with timestamp requests, which introduce artificial latency, throughput, and scalability bottlenecks.
>
>WARNING 2: Since we measure individual invocation timings with this level, we probably set ourselves up for (coordinated) omission. That means the hiccups in measurement can be hidden from timing measurement, and can introduce surprising results. For example, when we use timings to understand the benchmark throughput, the omitted timing measurement will result in lower aggregate time, and fictionally *larger* throughput.
>
>WARNING 3: In order to maintain the same sharing behavior as other Levels, we sometimes have to synchronize (arbitrage) the access to state objects. Other levels do this outside the measurement, but at this level, we have to synchronize on *critical path*, further offsetting the measurement.
>
>WARNING 4: Current implementation allows the helper method execution at this Level to overlap with the benchmark invocation itself in order to simplify arbitrage. That matters in multi-threaded benchmarks, when one worker thread executing benchmark method may observe other worker thread already calling `TearDown` for the same object.